File: PHP/DoublyCircularLinkedList.php
<?php


// to run this file, use this command: php PHP/DoublyCircularLinkedList.php
class DoublyNode {
    public object $data;
    
    public DoublyNode $prev;
    public DoublyNode $next;


    public function __construct(object $data) {
        $this->data = $data;
    }
}


class DoublyCircularLinkedList {    
    private DoublyNode $head;
    private DoublyNode $tail;

    private int $size;


    public function getHead(): DoublyNode {
        return $this->head;
    }


    public function getTail(): DoublyNode {
        return $this->tail;
    }
    

    public function getSize(): int {
        return $this->size;
    }


    public function addElementToStart(object $data): void {
        $newHeadNode = new DoublyNode($data);

        if ($this->size == 0) {
            $this->head = $newHeadNode;
            $this->tail = $newHeadNode;

            $newHeadNode->next = $newHeadNode;
            $newHeadNode->prev = $newHeadNode;
        }
        else {
            $this->head->prev = $newHeadNode;
            $newHeadNode->next = $this->head;
            $newHeadNode->prev = $this->tail;
            $this->tail->next = $newHeadNode;
            $this->head = $newHeadNode;
        }
        
        $this->size++;
    }


    public function deleteFirstElement(): void {
        if ($this->size == 0) {
            return;
        }

        $newHeadNode = $this->head->next;
        $newHeadNode->prev = $this->tail;
        $this->tail->next = $newHeadNode;
        $this->head = $newHeadNode;

        $this->size--;
    }


    public function addElementToEnd(object $data): void {
        $newTailNode = new DoublyNode($data);

        if ($this->size == 0) {
            $this->head = $newTailNode;
            $this->tail = $newTailNode;

            $newTailNode->next = $newTailNode;
            $newTailNode->prev = $newTailNode;
        }
        else {
            $this->tail->next = $newTailNode;
            $newTailNode->prev = $this->tail;
            $newTailNode->next = $this->head;
            $this->head->prev = $newTailNode;
            $this->tail = $newTailNode;
        }

        $this->size++;
    }


    public function deleteLastElement(): void {
        if ($this->size == 0) {
            return;
        }

        $newTailNode = $this->tail->prev;
        $newTailNode->next = $this->head;
        $this->head->prev = $newTailNode;
        $this->tail = $newTailNode;

        $this->size--;
    }
}

File: Python/DoublyCircularLinkedList.py
from typing import Any, Optional


#to run this file, use this command: python3 Python/DoublyCircularLinkedList.py
class DoublyNode:
    data: Any
    
    prev: Optional['DoublyNode']
    next: Optional['DoublyNode']


    def __init__(self, data):
        self.data = data


class DoublyCircularLinkedList:
    head: Optional['DoublyNode']
    tail: Optional['DoublyNode']

    size: int


    def get_head(self) -> Optional['DoublyNode']:
        return self.head


    def get_tail(self) -> Optional['DoublyNode']:
        return self.tail
    

    def get_size(self) -> int:
        return self.size


    def add_element_to_start(self, data: Any) -> None:
        new_head_node = DoublyNode(data)

        if self.size == 0:
            self.head = new_head_node
            self.tail = new_head_node

            new_head_node.next = new_head_node
            new_head_node.prev = new_head_node
        
        else:
            self.head.prev = new_head_node
            new_head_node.next = self.head
            new_head_node.prev = self.tail
            self.tail.next = new_head_node;
            self.head = new_head_node
        
        self.size+= 1


    def delete_first_element(self) -> None:
        if self.size == 0:
            return
        
        new_head_node = self.head.next
        new_head_node.prev = self.tail
        self.tail.next = new_head_node
        self.head = new_head_node

        self.size-= 1


    def add_element_to_end(self, data: Any) -> None:
        new_tail_node = DoublyNode(data)

        if self.size == 0:
            self.head = new_tail_node
            self.tail = new_tail_node

            new_tail_node.next = new_tail_node
            new_tail_node.prev = new_tail_node
        
        else:
            self.tail.next = new_tail_node
            new_tail_node.prev = self.tail
            new_tail_node.next = self.head
            self.head.prev = new_tail_node
            self.tail = new_tail_node

        self.size+= 1


    def delete_last_element(self) -> None:
        if self.size == 0:
            return
        
        new_tail_node = self.tail.prev
        new_tail_node.next = self.head
        self.head.prev = new_tail_node
        self.tail = new_tail_node

        self.size-= 1
